org.eclipse.xtext.util/src/org/eclipse/xtext/util/UriExtensions.java

```diff
@@ -8,35 +8,103 @@
  */
 package org.eclipse.xtext.util;
 
-import java.net.URISyntaxException;
+import java.util.Arrays;
 
 import org.eclipse.emf.common.util.URI;
 
+import com.google.inject.Singleton;
+
 /**
  * @author Christian Dietrich - Initial contribution and API
  * @since 2.14
  */
+@Singleton
 public class UriExtensions {
 	/**
-	 * Returns an URI with empty authority, if authority is absent (null) and has file scheme.
+	 * Converts the given absolute, encoded stringUri to an EMF URI. If the URI
+	 * is a {@link URI#isFile() file URI}, it is guaranteed that its
+	 * {@link URI#authority() authority} is not null.
 	 */
 	public URI toUri(String stringUri) {
-		try {
-			java.net.URI netUri = new java.net.URI(stringUri);
-			return toEmfUri(netUri);
-		} catch (URISyntaxException e) {
-			throw new RuntimeException(e);
+		URI result = URI.createURI(stringUri);
+		// Short-circuit the good case where the URI doesn't contain any unusual escape sequences
+		if (stringUri.contains("%")) {
+			return cleanupEncodedCharacters(result);
```

Comment:
It would even be faster if we do not need to reencode the string :)

I still does not quite get why it is so. Has it anything to do with Cyrillic and other characters as such? I tested with URIs like the ones we have (e.g.` ctx:/src/AAA/MBA/MBA$MASS%23STEX.SCRIPT%20PACKAGE`) and the recoding does not do anything visible. It is exactly the same afterwards.

Maybe it is overdoing it, but since we only allow ASCII characters on our files URIs, could we put maybe `stringUri.contains("%")` in a protected method that we could override to just return false?

Reply:
I'll extract a protected method.

Reply:
Done.